src/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -135,20 +135,41 @@ public IPAddress(byte[] address, long scopeid)
                 throw new ArgumentException(SR.dns_bad_ip_address, nameof(address));
             }
 
+            // Consider: Since scope is only valid for link-local and site-local
+            //           addresses we could implement some more robust checking here
+            if (scopeid < 0 || scopeid > 0x00000000FFFFFFFF)
+            {
+                throw new ArgumentOutOfRangeException(nameof(scopeid));
+            }
+
             _numbers = new ushort[NumberOfLabels];
 
             for (int i = 0; i < NumberOfLabels; i++)
             {
                 _numbers[i] = (ushort)(address[i * 2] * 256 + address[i * 2 + 1]);
             }
 
+            PrivateScopeId = (uint)scopeid;
+        }
+
+        internal unsafe IPAddress(byte* address, int addressLength, long scopeid)
```

Comment:
FYI: in places like this where I added a new ctor, it's the same as the ```byte[]``` versions, just using internal instead of public, using ```byte*``` instead of ```byte[]``` (and thus taking a length argument as well), and with asserts instead of exceptions to verify the address.